### ISSUE TITLE
Odd lenRight constraint and types mismatch

### ISSUE THREAD

**Issue by llelf** [ucsd-progsys/liquidhaskell#423]
Here’s a small example

```
module Main where

import Data.Either -- not used

foo :: [()] -> ()
foo xs = foo xs
```

Notice the types in the liquid output
```
 t.hs:7:14-15: Error: Liquid Type Mismatch
   Inferred type
     VV : {VV : [()] | VV == xs && len VV >= 0 && lenRight VV >= 0 && lenRight VV <= len VV}
    …  
   In Context
    …
     xs : {v : [()] | len v >= 0 && lenRight v >= 0 && lenRight v <= len v}
```

Whereas, the same code with commented out Data.Either import produces a reasonable output:
```
t.hs:7:14-15: Error: Liquid Type Mismatch
   Inferred type
     VV : {VV : [()] | VV == xs && len VV >= 0}
  
   not a subtype of Required type
     VV : {VV : [()] | len VV < len xs && len VV >= 0}
  
   In Context
     VV : {VV : [()] | VV == xs && len VV >= 0}
     xs : {v : [()] | len v >= 0}
```

**Comment by ranjitjhala** [comment]
Hmm, thanks for pointing this out! (Btw, it may be unclear but this is a
"termination" related error -- LH expects that that default list-valued
argument should get smaller,
hence is trying to prove that 'len VV < len xs' at the recursive call site.)

What's happening is that:

When you import Data.Either, LH eagerly imports the specifications from the
prelude:


https://github.com/ucsd-progsys/liquidhaskell/blob/master/include/Data/Either.spec#L5

and those facts (about the `lenRight` are getting added to the refinements.)

However, @nikivazou -- this is indeed a bug because the lenRight should
**only** apply to [Either a b] while here we have an incompatible `[()]`.
So the question is: why is the sort checker not pruning the measure
strengthening here?

Thanks again -- this is surprising, and will will fix it ASAP!


On Sat, Jul 25, 2015 at 1:43 PM, Antonio Nikishaev <notifications@github.com
> wrote:

> Here’s a small example
>
> module Main where
>
> import Data.Either -- not used
>
> foo :: [()] -> ()
> foo xs = foo xs
>
> Notice the types in the liquid output
>
>  t.hs:7:14-15: Error: Liquid Type Mismatch
>    Inferred type
>      VV : {VV : [()] | VV == xs && len VV >= 0 && lenRight VV >= 0 && lenRight VV <= len VV}
>     …
>    In Context
>     …
>      xs : {v : [()] | len v >= 0 && lenRight v >= 0 && lenRight v <= len v}
>
> Whereas, the same code with commented out Data.Either import produces a
> reasonable output:
>
> t.hs:7:14-15: Error: Liquid Type Mismatch
>    Inferred type
>      VV : {VV : [()] | VV == xs && len VV >= 0}
>
>    not a subtype of Required type
>      VV : {VV : [()] | len VV < len xs && len VV >= 0}
>
>    In Context
>      VV : {VV : [()] | VV == xs && len VV >= 0}
>      xs : {v : [()] | len v >= 0}
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ucsd-progsys/liquidhaskell/issues/423>.
>



-- 
Ranjit.

**Comment by gridaphobe** [comment]
The `lenRight` constraint is coming from your import of `Data.Either`, where we've defined `lenRight` as a measure that describes the number of `Right`s in a list (https://github.com/ucsd-progsys/liquidhaskell/blob/master/include/Data/Either.spec#L5-L7).

That being said, `lenRight` should *not* be appearing in the refinement in this case because it's ill-typed. @nikivazou shouldn't the sort-checker catch this (the mismatch between `[()]` and `[Either a b]`)?

**Comment by nikivazou** [comment]
You are right, sort checker should prune 'lenRight' out. I suspect there is
a bug when pruning and termination checker interact... Will look at it soon.

**Comment by nikivazou** [comment]
Also, @llelf if you want to turn the termination checker off you can run liquidHaskell with a `--no-termination` option, or use a 

```
{-@ Liquid --no-termination @-} 
```

Pragma in your code. Alternatively, 

```
{-@ lazy foo @-}
```

will deactivate termination checking for only the function `foo`.
